flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/input/MultiStateKeyIterator.java

```diff
@@ -31,65 +31,82 @@
+ *     href="https://bugs.openjdk.org/browse/JDK-8267359">https://bugs.openjdk.org/browse/JDK-8267359</a>
  * @param <K> Type of the key by which state is keyed.
  */
 @Internal
-public final class MultiStateKeyIterator<K> implements CloseableIterator<K> {
+public class MultiStateKeyIterator<K> implements CloseableIterator<K> {
     private final List<? extends StateDescriptor<?, ?>> descriptors;
 
     private final KeyedStateBackend<K> backend;
 
-    private final Iterator<K> internal;
+    private Iterator<? extends StateDescriptor<?, ?>> outerIter;
+    private Iterator<K> innerIter;
 
     private final CloseableRegistry registry;
 
     private K currentKey;
 
     public MultiStateKeyIterator(
             List<? extends StateDescriptor<?, ?>> descriptors, KeyedStateBackend<K> backend) {
+
+        outerIter = descriptors.iterator();
+        innerIter = null;
+
         this.descriptors = Preconditions.checkNotNull(descriptors);
         this.backend = Preconditions.checkNotNull(backend);
 
         this.registry = new CloseableRegistry();
-        this.internal =
-                descriptors.stream()
-                        .map(
-                                descriptor ->
-                                        backend.getKeys(
-                                                descriptor.getName(), VoidNamespace.INSTANCE))
-                        .peek(
-                                stream -> {
-                                    try {
-                                        registry.registerCloseable(stream::close);
-                                    } catch (IOException e) {
-                                        throw new RuntimeException(
-                                                "Failed to read keys from configured StateBackend",
-                                                e);
-                                    }
-                                })
-                        .flatMap(stream -> stream)
-                        .iterator();
+    }
+
+    @Override
+    public void close() throws Exception {
+        registry.close();
     }
 
     @Override
     public boolean hasNext() {
-        return internal.hasNext();
+        if ((innerIter == null || !innerIter.hasNext()) && outerIter.hasNext()) {
+            StateDescriptor<?, ?> descriptor = outerIter.next();
+            Stream<K> stream = backend.getKeys(descriptor.getName(), VoidNamespace.INSTANCE);
+            innerIter = stream.iterator();
+            try {
+                registry.registerCloseable(stream::close);
+            } catch (IOException e) {
+                throw new RuntimeException("Failed to read keys from configured StateBackend", e);
+            }
+        }
+        if (innerIter == null) {
+            return false;
+        }
+        return innerIter.hasNext();
     }
 
     @Override
     public K next() {
-        currentKey = internal.next();
-        return currentKey;
+        if (!this.hasNext()) {
```

Comment:
It's not necessary due to contract of `Iterator`.
This will increase unnecessary cost when using iterator.

Reply:
`this.hasNext()` is the only place that initializes the `innerIter` and therefore is neccessary for the case the caller hasn't called `hasNext()` before.